### PR TITLE
Expanding HLT Configuration Proxy Support and Adding Tunnel Support

### PR DESCRIPTION
**Pull request by Sam-Harper**
#### PR description:

Confdb access is restricted to within the CERN network and thus for users outside CERN they must connect via tunnels. 

Currently hltGetConfiguration and hltConfigFromDB allow the use of a socks proxy to allow users outside CERN to download menus. However the IPs need to be harcoded due to DNS issues (an issue with Java which as far as we are aware is unsolvable). 

Only the offline DB had its IPs hardcoded, we now add IPs for the online (p5 only) and online mirror (accessable outside p5) so users can also use a proxy for these dbs. 

Finally to have a work around in the future, support for direct tunnels where a user specifically forwards a given port on their machine to a db has been added so we can always fall back on this.

Aim to back port to 12_4, 12_5, 12_6 and ~~10_6~~ (I believe these are all the releases users are using for analysis). 

edit: probably not 10_6, it hasn't received the other updates necessary unless there is a strong push for it.
#### PR validation:

Configs from online/ online-mirror and offline were downloaded at RAL using both direct tunnels and a proxy server.

### CODE REVIEW

**Comment by Sam-Harper** [comment]
please test

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 1]
```diff
@@ -109,6 +119,16 @@ def __init__(self, version = 'v3', database = 'run3', url = None, verbose = Fals
...
 
+        #this sets the host to localhost
+        if self.tunnel:
+            temp_connect = list(self.connect)
+            host_index = temp_connect.index('-h')
+            temp_connect[host_index+1] = "localhost"
+            self.connect = tuple(temp_connect)
+            self.tunnel_connect_args = ('--dbport',self.tunnelPort)
+        else:
+            self.tunnel_connect_args
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:130]:
```suggestion
            self.tunnel_connect_args = ()
```
Not sure if this works without the `= ()`.

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 2]
```diff
@@ -288,7 +308,20 @@ def main():
     if '--dbproxyport' in args:
         proxy_port = args.pop(args.index('--dbproxyport')+1)
         args.remove('--dbproxyport')
-        
+
+    tunnel = False
+    tunnel_port = "10121"
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:313]: Note to self (not for this PR): we should find a way to set the default values only once (e.g. in `options.py`), without having to hard-code more than once.

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 3]
```diff
@@ -288,7 +308,20 @@ def main():
...
+    if '--dbtunnel' in args:
+        tunnel = True
+        args.remove('--dbtunnel')
+
+    if '--dbtunnelport' in args:
+        tunnel_port =  args.pop(args.index('--dbtunnelport')+1)
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:319]:
```suggestion
        tunnel_port = args.pop(args.index('--dbtunnelport')+1)
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 4]
```diff
@@ -288,7 +308,20 @@ def main():
...
+        args.remove('--dbtunnelport')
+
+    if tunnel and proxy:
+        sys.stderr.write( 'ERROR: conflicting connection specifications, "--dbtunnel" and "--proxy" are mutually exclusive options\n')
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:323]:
```suggestion
        sys.stderr.write( 'ERROR: conflicting connection specifications, "--dbtunnel" and "--dbproxy" are mutually exclusive options\n' )
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 5]
```diff
@@ -315,7 +348,9 @@ def main():
             sys.exit(1)
 
     converter = OfflineConverter(version = version, database = db, verbose = verbose,
-                                 proxy = proxy, proxyHost = proxy_host, proxyPort=proxy_port)
+                                 proxy = proxy, proxyHost = proxy_host, proxyPort=proxy_port,
+                                 tunnel=tunnel,tunnelPort=tunnel_port
+    )
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:353]:
```suggestion
                                 tunnel = tunnel, tunnelPort = tunnel_port)
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 6]
```diff
@@ -46,7 +46,9 @@ class OfflineConverter:
     ips_for_proxy = {
         'cmsr1-s.cern.ch' : '10.116.96.89',
         'cmsr2-s.cern.ch' : '10.116.96.139',
-        'cmsr3-s.cern.ch' : '10.116.96.105'
+        'cmsr3-s.cern.ch' : '10.116.96.105',
+        'cmsonr1-adg1-s.cern.ch' : '10.116.96.109',
+        'cmsonr1-s.cms': '10.176.84.78'
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:51]:
```suggestion
        'cmsonr1-s.cms' : '10.176.84.78'
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 7]
```diff
@@ -70,7 +72,9 @@ def CheckTempDirectory(dir):
 
 
     def __init__(self, version = 'v3', database = 'run3', url = None, verbose = False,
-                 proxy = False, proxyHost = 'localhost', proxyPort = '8080'): 
+                 proxy = False, proxyHost = 'localhost', proxyPort = '8080',
+                 tunnel = False, tunnelPort = '10121'
+                ):
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:77]:
```suggestion
                 tunnel = False, tunnelPort = '10121'):
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 8]
```diff
@@ -83,6 +87,12 @@ def __init__(self, version = 'v3', database = 'run3', url = None, verbose = Fals
         self.proxy = proxy
         self.proxyHost = proxyHost
         self.proxyPort = proxyPort
+        self.tunnel = tunnel
+        self.tunnelPort = tunnelPort
+
+        if self.proxy and self.tunnel:
+            sys.stderr.write( "ERROR: proxy and tunnel options can not both be true ")
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:94]:
```suggestion
            sys.stderr.write( "ERROR: proxy and tunnel options can not both be true" )
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 9]
```diff
@@ -109,6 +119,16 @@ def __init__(self, version = 'v3', database = 'run3', url = None, verbose = Fals
         else:
             self.proxy_connect_args = ()
 
+        #this sets the host to localhost
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:122]:
```suggestion
        # this sets the host to localhost
```

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 10]
```diff
@@ -109,6 +119,16 @@ def __init__(self, version = 'v3', database = 'run3', url = None, verbose = Fals
...
+        if self.tunnel:
+            temp_connect = list(self.connect)
+            host_index = temp_connect.index('-h')
+            temp_connect[host_index+1] = "localhost"
+            self.connect = tuple(temp_connect)
+            self.tunnel_connect_args = ('--dbport',self.tunnelPort)
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:128]:
```suggestion
            self.tunnel_connect_args = ('--dbport', self.tunnelPort)
```

**Review by missirol — COMMENTED** [review]
Thanks for this update! One change is maybe needed, the other comments are optional (cosmetics).

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py (hunk 1 above)
Comment (Sam-Harper) [reply HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:130]: ah thanks!

**Comment by Sam-Harper** [comment]
Thanks Marino for the review!

**Comment by missirol** [comment]
Thanks, Sam. I hate to be picky, but would you mind squashing things in 1 single commit?

**Comment by Sam-Harper** [comment]
no problem its better that way, done

**Comment by Sam-Harper** [comment]
please test

**Comment by missirol** [comment]
I tested the db proxy/tunnel options inside the CERN network, and found no problems [*] [**]. On Monday, I can test this outside the CERN network (the T3 I have access to is under maintenance till then), and can then sign this PR.

I noticed a couple of things that could be added to this PR already, if possible. They are in the following diff.
```diff
diff --git a/HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py b/HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py
index 2b28e7e2624..a89a5aa0b88 100644
--- a/HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py
+++ b/HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py
@@ -242,6 +242,14 @@ def help():
         --modules <p1[,p2]>         (include modules, referenced or not!)
         --blocks <m1::p1[,p2][,m2]> (generate parameter blocks)
 
+        Options to connect to target db via SOCKS proxy, or direct tunnel:
+          [the options --dbproxy and --dbtunnel are mutually exclusive]
+        --dbproxy                   (use a SOCKS proxy to connect outside CERN network [default: False])
+        --dbproxyport <port>        (port of the SOCKS proxy [default: 8080])
+        --dbproxyhost <hostname>    (host of the SOCKS proxy [default: "localhost"])
+        --dbtunnel                  (use direct tunnel to connect outside CERN network [default: False])
+        --dbtunnelport <port>       (port when using a direct tunnel on localhost [default: 10121])
+
         --verbose                   (print additional details)
 """)
 
@@ -294,8 +302,8 @@ def main():
         version = 'v3-test'
         db      = 'dev'
         args.remove('--v3-test')
-    
-    proxy=False
+
+    proxy = False
     proxy_host = "localhost"
     proxy_port = "8080"
     if '--dbproxy' in args:
@@ -347,11 +355,11 @@ def main():
             sys.exit(1)
 
     converter = OfflineConverter(version = version, database = db, verbose = verbose,
-                                 proxy = proxy, proxyHost = proxy_host, proxyPort=proxy_port,
+                                 proxy = proxy, proxyHost = proxy_host, proxyPort = proxy_port,
                                  tunnel = tunnel, tunnelPort = tunnel_port)
     out, err = converter.query( * args )
     if 'ERROR' in err:
-        sys.stderr.write( "%s: error while retriving the HLT menu\n\n%s\n\n" % (sys.argv[0], err) )
+        sys.stderr.write( "%s: error while retrieving the HLT menu\n\n%s\n\n" % (sys.argv[0], err) )
         sys.exit(1)
     else:
         sys.stdout.write( out )
diff --git a/HLTrigger/Configuration/scripts/hltGetConfiguration b/HLTrigger/Configuration/scripts/hltGetConfiguration
index 587cfe25cbd..1a0a6aa338b 100755
--- a/HLTrigger/Configuration/scripts/hltGetConfiguration
+++ b/HLTrigger/Configuration/scripts/hltGetConfiguration
@@ -255,14 +255,14 @@ parser.add_argument('--help',
 # parse command line arguments and options
 config = parser.parse_args(namespace = options.HLTProcessOptions())
 
-# do not include db-proxy options in 1st-line comment
+# do not include db-proxy/tunnel options in 1st-line comment
 cmdArgs, skipNext = [], False
 for cmdArg in sys.argv:
   if skipNext:
     skipNext = False
     continue
-  if cmdArg.startswith('--dbproxy'):
-    if cmdArg.startswith('--dbproxyh') or cmdArg.startswith('--dbproxyp'):
+  if cmdArg.startswith('--dbproxy') or cmdArg.startswith('--dbtunnel'):
+    if cmdArg.startswith('--dbproxyh') or cmdArg.startswith('--dbproxyp') or cmdArg.startswith('--dbtunnelp'):
       skipNext = '=' not in cmdArg
     continue
   cmdArgs += [cmdArg]
```

[*] In my case, the direct tunnel works as long as I use port=`10121` (as opposed to other port numbers). I was wondering if that's expected. Are the other ports 'blocked'?

[**] There is an edge case in `hltConfigFromDB` that could be fixed (having to do with using `=`), as this won't work
```
hltConfigFromDB --adg --configName /cdaq/physics/Run2022HI/v1.1.0/HLT/V11 --dbtunnel --dbtunnelport=10121
```
while the following does (courtesy of `argparse`)
```
hltGetConfiguration adg:/cdaq/physics/Run2022HI/v1.1.0/HLT/V11 --dbtunnel --dbtunnelport=10121
```
On the other hand, this 'limitation' already exists for `dbproxy*`, so I think it'd be okay to leave it for a future PR.

**Comment by Sam-Harper** [comment]
```
[**] There is an edge case in hltConfigFromDB that could be fixed (having to do with using =), as this won't work

hltConfigFromDB --adg --configName /cdaq/physics/Run2022HI/v1.1.0/HLT/V11 --dbtunnel --dbtunnelport=10121
while the following does (courtesy of argparse)

hltGetConfiguration adg:/cdaq/physics/Run2022HI/v1.1.0/HLT/V11 --dbtunnel --dbtunnelport=10121
```

Thats sadly just a feature of hltConfigFromDB, it works as `--option value` not `--option=value`. Changing would apply  to all arguements, eg `--configName  /cdaq/physics/Run2022HI/v1.1.0/HLT/V11` to `--configName=/cdaq/physics/Run2022HI/v1.1.0/HLT/V11` . So as you say, another time. 

```
[*] In my case, the direct tunnel works as long as I use port=10121 (as opposed to other port numbers). I was wondering if that's expected. Are the other ports 'blocked'?
```

hmm i tested with 10122 using `-L 10122:<dbserver>:10121` and it worked? 

```I noticed a couple of things that could be added to this PR already, if possible. They are in the following diff.```
Sure will impliment it monday

**Comment by missirol** [comment]
Tested outside the CERN network, and works as expected.

**Comment by Sam-Harper** [comment]
new PR: https://github.com/cms-sw/cmssw/pull/40454

**Pull request closed without merging**